Objects/regions.c

```diff
@@ -670,8 +720,277 @@ PyObject* _Py_MakeImmutable(PyObject* obj)
+        .region = region,
+    };
+    if (info.pending == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    if (stack_push(info.pending, obj)) {
+        stack_free(info.pending);
+        return PyErr_NoMemory();
+    }
+
+    regionmetadata *region_data = _Py_CAST(regionmetadata *, region);
+    while (!stack_empty(info.pending)) {
+        PyObject *item = stack_pop(info.pending);
+        PyTypeObject *type = Py_TYPE(item);
+        regionmetadata *item_region = _Py_CAST(regionmetadata *, Py_REGION(item));
+
+        if (Py_REGION(item) == region) {
+            // -1 for the refernce we just followed
+            region_data->lrc -= 1;
+            continue;
+        }
+
+        if (IS_DEFAULT_REGION(Py_REGION(item))) {
+            // Add reference to the object,
+            // minus one for the reference we just followed
+            region_data->lrc += item->ob_refcnt - 1;
+            Py_SET_REGION(item, flagged_region);
+            item_region = region_data;
+
+            if (PyFunction_Check(item)) {
+                regionerror err = {.src = _PyObject_CAST(region_data->bridge),
+                    .tgt = item, .id = ERR_WIP_FUNCTIONS };
+                if ((handle_error)(&err, handle_error_data)) {
+                    continue;
+                } else {
+                    goto fail;
+                }
+            } else {
+                traverseproc traverse = type->tp_traverse;
+                if (traverse != NULL) {
+                    if (traverse(item, (visitproc)_add_to_region_visit, &info)) {
+                        stack_free(info.pending);
+                        return NULL;
+                    }
+                }
+            }
+
+            PyObject* type_op = _PyObject_CAST(Py_TYPE(item));
+            if (Py_REGION(type_op) != region && !_Py_IsImmutable(type_op)) {
+                if (stack_push(info.pending, type_op))
+                {
+                    stack_free(info.pending);
+                    return PyErr_NoMemory();
+                }
+            }
+
+            continue;
+        }
+
+        if (_Py_IsImmutable(item)) {
+            continue;
+        }
+
+        // C wrappers can propergate through the entire system and draw
+        // in a lot of unwanted objects. Since c wrappers don't have mutable
+        // data, we just make it immutable and have the immutability impl
+        // handle it. We then have an edge from our region to an immutable
+        // object which is again valid.
+        if(is_c_wrapper(item)) {
+            _Py_MakeImmutable(item);
+            continue;
+        }
+
+        if (!is_bridge_object(item)) {
+            // FIXME: This uses the bridge object for now, since we
+            // currently don't store the reference source. We need
+            // to figure out where we can get the source from since the
+            // stack only holds one object. Or maybe we don't even need
+            // the source object?
```

Comment:
I wonder if this means that we need to put more of this code into the visitproc?  We could put the information about the object currently being traversed into the `info` passed to visitproc, and then we would know that when we get to this code.  Otherwise, we could put it into the stack, but then the stack becomes twice the size, which seems bad for better error messages.